### ISSUE TITLE
Failed to mesh with oriented surface patches

### ISSUE THREAD

**Issue by wangbowenll** [CGAL/cgal#3897]
## Issue Details
Dear researcher
I use CGAL::orientation(p, q, r, s) to check for one triangle (p,q,r) of the surface, and s a point inside the chosen subdomain, if the orienation is positive or negative. 
Then i construct the std::pair<subdomainindexA,subdomainindexB> according to the orientation result.A is in positive side and B is in negative side.
I checked model, it is same segment,and same point. but it crashes and throw ":Expression : ! is_infinite(c->vertex(i)) "

What am i missing?
Attatchment is new dataset about the sedmentary model
[Sedimentary model.zip](https://github.com/CGAL/cgal/files/3141127/Sedimentary.model.zip)

<img src="https://user-images.githubusercontent.com/43980713/57126759-ad4a8000-6dc0-11e9-8af1-73b718672ea5.jpg" width="350">

## Source Code

```
#define CGAL_CONCURRENT_MESH_3
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>

#include <CGAL/Mesh_triangulation_3.h>
#include <CGAL/Mesh_complex_3_in_triangulation_3.h>
#include <CGAL/Mesh_criteria_3.h>

#include <CGAL/Polyhedral_complex_mesh_domain_3.h>
#include <CGAL/make_mesh_3.h>

//#include <CGAL/Polyhedron_3.h>

#include <cstdlib>

// Domain

typedef CGAL::Exact_predicates_inexact_constructions_kernel K;
typedef CGAL::Mesh_polyhedron_3<K>::type Polyhedron;
typedef CGAL::Polyhedral_complex_mesh_domain_3<K,Polyhedron> Mesh_domain;

typedef CGAL::Polyhedron_3<K>                   Polyhedron2;


typedef Polyhedron2::Facet_iterator                   Facet_iterator;
typedef Polyhedron2::Halfedge_around_facet_circulator  Halfedge_facet_circulator;
typedef Polyhedron2::Vertex_iterator        Vertex_iterator;
typedef K::Point_3                              Point_3;



#ifdef CGAL_CONCURRENT_MESH_3
typedef CGAL::Parallel_tag Concurrency_tag;
#else
typedef CGAL::Sequential_tag Concurrency_tag;
#endif

// Triangulation
typedef CGAL::Mesh_triangulation_3<Mesh_domain,CGAL::Default,Concurrency_tag>::type Tr;

typedef CGAL::Mesh_complex_3_in_triangulation_3<
  Tr,Mesh_domain::Corner_index,Mesh_domain::Curve_index> C3t3;

// Criteria
typedef CGAL::Mesh_criteria_3<Tr> Mesh_criteria;

// To avoid verbose function and named parameters call
using namespace CGAL::parameters;

const char* const filenames[] = {	
	"data/4/0011.off",
	"data/4/12.off",
	"data/4/20.off",	
	"data/4/23.off",	
 };

const std::pair<int, int> incident_subdomains[] = {
	std::make_pair(0, 1),
	std::make_pair(1, 2),
	std::make_pair(0, 2),
	std::make_pair(2, 0),	 	   	
};

int main()
{	
  const std::size_t nb_patches = sizeof(filenames) / sizeof(const char*);

  CGAL_assertion(sizeof(incident_subdomains) ==
                 nb_patches * sizeof(std::pair<int, int>));

  std::vector<Polyhedron> patches(nb_patches);
  std::vector<Polyhedron2> patch(nb_patches);

  for(std::size_t i = 0; i < nb_patches; ++i) {
    std::ifstream input(filenames[i]);
    if(!(input >> patches[i])) {
      std::cerr << "Error reading " << filenames[i] << " as a polyhedron!\n";
      return EXIT_FAILURE;
	  }
	else
	{
		std::cout << "reading " << filenames[i] << " as a polyhedron!\n";
	}
    }
  for (std::size_t i = 0; i < nb_patches; ++i) {
	  std::ifstream input2(filenames[i]);
	  if (!(input2 >> patch[i])) {
		  std::cerr << "Error reading " << filenames[i] << " as a polyhedron!\n";
		  return EXIT_FAILURE;
	  }
	  else
	  {
		  std::cout << "reading " << filenames[i] << " as a polyhedron!\n";
	  }
  }

  // Create domain
  Mesh_domain domain(patches.begin(), patches.end(),
                     incident_subdomains, incident_subdomains+nb_patches);

  
  domain.detect_features(); //includes detection of borders

  Facet_iterator i = patch[0].facets_begin();
  Halfedge_facet_circulator j = i->facet_begin();

  std::vector<Point_3> point0(4);
  Point_3 p(-222.763, -397.364, 200);  // p   outside subdomain1
  Point_3 q(-222.763, -397.364, 145);  // q  inside subdomain1
  int k = 0;
  do{
	  std::cout << j->vertex()->point() << "\n";
	  std::cout << j->vertex()->point() << "\n";
	  point0[k] = j->vertex()->point();
	  std::cout << point0[k]<< "\n";
	  k++;
  } while (++j != i->facet_begin());

  switch (CGAL::orientation(point0[0], point0[1], point0[2], p)) {
  case  CGAL::POSITIVE:
		  std::cout << "are positive\n";
		  break;
  case   CGAL::NEGATIVE:
		 std::cout << "are negative\n";
		break;
  }    //positive
  switch (CGAL::orientation(point0[0], point0[1], point0[2], q)) {
  case  CGAL::POSITIVE:
	  std::cout << "are positive\n";
	  break;
  case   CGAL::NEGATIVE:
	  std::cout << "are negative\n";
	  break;
  }  //negative

// check subdomain2
   i = patch[1].facets_begin();
   j = i->facet_begin();
  std::vector<Point_3> point1(4);

   k = 0;
  do {
	  std::cout << j->vertex()->point() << "\n";
	  std::cout << j->vertex()->point() << "\n";
	  point1[k] = j->vertex()->point();
	  std::cout << point1[k] << "\n";
	  k++;
  } while (++j != i->facet_begin());

  switch (CGAL::orientation(point1[0], point1[1], point1[2], q)) {
  case  CGAL::POSITIVE:
	  std::cout << "are positive\n";
	  break;
  case   CGAL::NEGATIVE:
	  std::cout << "are negative\n";
	  break;
  }  // positive


  i = patch[2].facets_begin();
  j = i->facet_begin();
  std::vector<Point_3> point2(4);
  Point_3 r(-32.6315, -224.839, 27); // r inside  subdomain 2
  Point_3 s(-32.6315, -224.839, 20);// s inside sudomain 2

  k = 0;
  do {
	  std::cout << j->vertex()->point() << "\n";
	  std::cout << j->vertex()->point() << "\n";
	  point2[k] = j->vertex()->point();
	  std::cout << point2[k] << "\n";
	  k++;
  } while (++j != i->facet_begin());

  switch (CGAL::orientation(point2[0], point2[1], point2[2], r)) {
  case  CGAL::POSITIVE:
	  std::cout << "are positive\n";
	  break;
  case   CGAL::NEGATIVE:
	  std::cout << "are negative\n";
	  break;
  } //negative


  i = patch[3].facets_begin();
  j = i->facet_begin();
  std::vector<Point_3> point3(4);
  k = 0;
  do {
	  std::cout << j->vertex()->point() << "\n";
	  std::cout << j->vertex()->point() << "\n";
	  point3[k] = j->vertex()->point();
	  std::cout << point3[k] << "\n";
	  k++;
  } while (++j != i->facet_begin());

  switch (CGAL::orientation(point3[0], point3[1], point3[2], r)) {
  case  CGAL::POSITIVE:
	  std::cout << "are positive\n";
	  break;
  case   CGAL::NEGATIVE:
	  std::cout << "are negative\n";
	  break;
  }//positive

  switch (CGAL::orientation(point3[0], point3[1], point3[2], s)) {
  case  CGAL::POSITIVE:
	  std::cout << "are positive\n";
	  break;
  case   CGAL::NEGATIVE:
	  std::cout << "are negative\n";
	  break;
  } //negative

  std::cout << "detect feature is done\n" ;

  // Mesh criteria
  Mesh_criteria criteria(edge_size = 5,
	  facet_angle = 25, facet_size = 5, facet_distance = 3,
	  cell_radius_edge_ratio = 2, cell_size =7);

  // Mesh generation
  C3t3 c3t3 = CGAL::make_mesh_3<C3t3>(domain, criteria,
	  no_exude(),
	  perturb(sliver_bound = 10, time_limit = 15));
 
  // Output
  std::ofstream medit_file("remadis.mesh");
  c3t3.output_to_medit(medit_file,false,true);

  return EXIT_SUCCESS;
}
```

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits):windows 64 bits
* Compiler:vs 2017 community and vc14.1 (in cmake visual studio 15 2017 win 64)
* Release or debug mode:debug mode
* Specific flags used (if any):
* CGAL version:CGAL 4.13
* Boost version:1.67
* Other libraries versions if used (Eigen, TBB, etc.):TBB +Engin

**Comment by lrineau** [comment]
I assign this issue to @janetournois, with whom you started discussing in your other issues.

Jane, let me know if you need me on that issue.

**Comment by janetournois** [comment]
I can reproduce the assertion. I will have a look at it, thank you!

**Comment by janetournois** [comment]
Hi @wangbowenll 

I am testing your data, and I am meeting the same issue as before : the sub-polyhedra are not exactly conformal, so the subdomains are not watertight, which is a requirement of Mesh_3.

Here is the pipeline I am using to check this :
- use [`CGAL::copy_face_graph()`](https://doc.cgal.org/latest/BGL/group__PkgBGLHelperFct.html#ga2d3eb044bfe42eab566779ce32216c11) to join two adjacent polyhedra in one
- use [`CGAL::Polygon_mesh_processing::stitch_borders()`](https://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__repairing__grp.html#ga15f047f56af2fee3e491120db4dc69a3) to stitch together the boundaries that are equal

In the case of 20.off and 23.off of your input files, this procedure gives a polyhedron with two connected components, because borders do not exactly match.

**Comment by janetournois** [comment]
There will be a boundary snapping function in the CGAL master branch soon, that is already available in the pull-request #3266

**Comment by wangbowenll** [comment]
> In the case of 20.off and 23.off of your input files, this procedure gives a polyhedron with two connected components, because borders do not exactly match, so we cannot expect Mesh_3 to terminate properly.

Hi janetournois 
@janetournois  Your mean the 20.off and 23.off donnot exactly match. however, when i change the pair"std::make_pair(0, 1),
	std::make_pair(1, 2),
	std::make_pair(0, 2),
	std::make_pair(2, 0),"
to  "std::make_pair(1, 0),
	std::make_pair(2, 1),
	std::make_pair(2, 0),
	std::make_pair(0, 2),"
the program can  correctly terminate. so i think the off should be right. the issue should be the pair of "<sundomain1,subdomain2>"
